clients/src/main/java/org/apache/kafka/common/metrics/stats/TokenBucket.java

```diff
@@ -0,0 +1,150 @@
+ * limitations under the License.
+ */
+package org.apache.kafka.common.metrics.stats;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.kafka.common.metrics.MetricConfig;
+
+public class TokenBucket extends SampledStat {
+
+    private final TimeUnit unit;
+
+    public TokenBucket() {
+        this(TimeUnit.SECONDS);
+    }
+
+    public TokenBucket(TimeUnit unit) {
+        super(0);
+        this.unit = unit;
+    }
+
+    @Override
+    public void record(MetricConfig config, double value, long timeMs) {
+        if (value < 0) {
+            unrecord(config, -value, timeMs);
+            return;
+        }
+
+        final double quota = quota(config);
+        final long firstTimeWindowMs = firstTimeWindowMs(config, timeMs);
+
+        // Get current sample or create one if empty
+        Sample sample = current(firstTimeWindowMs);
+
+        // Verify that the current sample was not reinitialized. If it was the case,
+        // restart from the first time window
+        if (sample.eventCount == 0) {
+            sample.reset(firstTimeWindowMs);
+        }
+
+        // Add the value to the current sample
+        sample.value += value;
+        sample.eventCount += 1;
+
+        // If current sample is completed AND a new one can be created,
+        // create one and spill over the amount above the quota to the
+        // new sample. Repeat until either the sample is not complete
+        // or no new sample can be created.
+        while (sample.isComplete(timeMs, config)) {
+            double extra = sample.value - quota;
+            sample.value = quota;
+
+            sample = advance(config, sample.lastWindowMs + config.timeWindowMs());
+            sample.value = extra;
+            sample.eventCount += 1;
+        }
+    }
+
+    private void unrecord(MetricConfig config, double value, long timeMs) {
+        final double quota = quota(config);
+        final long firstTimeWindowMs = firstTimeWindowMs(config, timeMs);
+
+        // Rewind
+        while (value > 0) {
+            // Get current sample or create one if empty
+            Sample sample = current(firstTimeWindowMs);
+
+            // If the current sample has been purged, we can't unrecord anything
+            if (sample.eventCount == 0) {
+                return;
+            }
+
+            if (value <= sample.value) {
+                sample.value -= value;
+                return;
+            }
+
+            sample.reset(timeMs);
+            value -= quota;
+
+            this.current = this.current - 1;
+            if (this.current <= 0)
+                this.current = this.samples.size() - 1;
+        }
+    }
+
+    private static double quota(MetricConfig config) {
+        if (config.quota() == null)
+            throw new IllegalStateException("TokenBucket can't be used without a quota.");
+
+        return config.quota().bound();
```

Comment:
I think this should return `config.quota().bound() * convert(config.timeWindowMs())`, because we are using the value returned from this method to compare to a value in a sample window. Also could be useful to add one unit test where MetricConfig.timeWindowMs() is not 1 second.

Reply:
and maybe consider renaming it to `sampleQuota` or `sampleWindowQuota` or something that hints that this method returns quota over a sample time window?

Reply:
You're absolutely right.